CONTRIBUTING.md

```diff
@@ -173,6 +173,47 @@ The `legacy_foo` arg was removed
 :::
 ```
 
+## Style and idioms
+
+For the most part, we just accept whatever the code formatters do, so there
+isn't much style to enforce.
+
+Some miscellanous style, idioms, and conventions we have are:
+
+### Markdown/Sphinx Style
+
+* Use colons for prose sections of text, e.g. `:::{note}`, not backticks.
+* Use backticks for code blocks.
+* Max line length: 100.
+
+### BUILD/bzl Style
+
+* When a macro generates public targets, use a dot (`.`) to separate the
+  user-provided name from the generted name. e.g. `foo(name="x")` generates
+  `x.test`. The `.` is our convention to communicate that it's a generated
+  target, and thus one should look for `name="x"` when searching for the
+  definition.
+* The different build phases and interface/implementation pieces of rules should
+  be in different files. Stated another way: loading code for thing shouldn't
+  incur loading code that isn't relevant.
+  * Providers should be in their own files. This allows implementing a custom
+    rule that implements the provider without loading an existing
+    implementation.
+  * The repository phase shouldn't `load()` code that is only relevant to the
+    loading-phase or subsequent phases. Vice-versa, loading-phase code should
+    not load any repository-phase code.
+  * Generally, one rule per file. The goal is that defining an e.g. library
+    shouldn't incur loading all the code for binaries, tests, packaging, etc.
+* The public access point for anything should be a separate file that re-exports
+  the specific symbols that are public. This ensures our public API is well
+  defined and prevents accidentally exposing a package-private symbol as a
+  public symbol.
+* Repository rules should have name ending in `_repo`. This helps distinguish
+  them from regular rules.
```

Comment:
hmmm. What do we do with existing `whl_library`, etc? I'm `+0` for the change though, might be good to add a cleanup.

However, I would like to have a list of permited names. For example I wanted to later create a `whl_archive` repo rule (similar to `http_archive`) that would be a starlark impl of extracting a `whl` and making it usable. I guess it would have to become `whl_archive_repo`?

Reply:
We can just leave what's out of compliance as it is until someone really wants to scratch the itch.

I'll give http_archive a _slight_ pass because its "http" portion makes it pretty clear it's going to be doing something external.

Is `whl_archive` a repository rule that extracts a wheel, or is it a regular rule that accepts/produces a `.whl` archive as input/output?

I'd love to rename whl_library. It's `_library` suffix has always given me cognitive dissonance because every other `xxx_library` I see is a regular rule.  I would take whl_archive over whl_library 100%.

Reply:
OK, SGTM. Having the `repo` makes sense though. The API of `rules_python` should be intuitive and having this explained somewhere in the docs would be good. Maybe such conventions belong not only in the style (which is for contributors) but in the API reference index, i.e. we should tell our users how the API is roughly organized.